test/e2e/e2e_test.go

```diff
@@ -0,0 +1,562 @@
+		originalLength := len(input)
+		for i := 1; i <= len(input); i++ { // this is really brute force but it'll do
+			var post post
+			err := json.Unmarshal(input[:i], &post)
+			if err == nil {
+				posts = append(posts, post)
+				input = input[i:]
+				break
+			}
+		}
+		if len(input) == originalLength {
+			return nil, errors.Errorf("couldn't parse full input, %d characters left, %d originally", len(input), originalLength)
+		}
+	}
+	return posts, nil
+}
+
+func deleteS3Object(bucket, key string) error {
+	cfg, err := config.LoadDefaultConfig(context.TODO())
+	if err != nil {
+		return err
+	}
+
+	ctx, cancelFunc := context.WithTimeout(context.Background(), time.Second*5)
+	defer cancelFunc()
+
+	client := s3.NewFromConfig(cfg)
+	_, err = client.DeleteObject(ctx, &s3.DeleteObjectInput{
+		Bucket: &bucket,
+		Key:    &key,
+	})
+
+	return err
+}
+
+func validatedEnvironment() (*environment, error) {
+	s3URL := os.Getenv("AWAT_E2E_BUCKET")
+	if s3URL == "" {
+		return nil, errors.New("provided bucket name must not be empty; set AWAT_E2E_BUCKET")
+	}
+
+	awat := os.Getenv("AWAT_E2E_URL")
+	if awat == "" {
+		return nil, errors.New("provided AWAT URL must not be empty; set AWAT_E2E_URL")
+	}
+
+	domain := os.Getenv("AWAT_E2E_INSTALLATION_DOMAIN")
+
+	if domain == "" {
+		return nil, errors.New("provided target Installation DNS space must not be empty; set AWAT_E2E_INSTALLATION_DOMAIN to e.g. .dev.cloud.mattermost.com")
+	}
+
+	provisionerURL := os.Getenv("AWAT_E2E_PROVISIONER_URL")
+	if provisionerURL == "" {
+		return nil, errors.New("provided Provisioner URL must not be empty; set AWAT_E2E_PROVISIONER_URL")
+	}
+
+	err := ensureArtifactInBucket(s3URL)
+	if err != nil {
+		return nil, errors.Wrap(err, "provided bucket was not valid")
+	}
+
+	return &environment{
+		awatURL:     awat,
+		bucket:      s3URL,
+		provisioner: cloud.NewClient(provisionerURL),
+		testDomain:  domain,
+	}, nil
+}
+
+func uploadTestArtifact(bucketName string) error {
+	cfg, err := config.LoadDefaultConfig(context.TODO())
+	if err != nil {
+		return err
+	}
+
+	client := s3.NewFromConfig(cfg)
+	archive, err := os.Open(slackArchive)
+	defer archive.Close()
+	if err != nil {
+		return err
+	}
+	keyName := strings.TrimPrefix(archive.Name(), "../") // forgive me oh Lord for the sins I have committed
+
+	params := &s3.PutObjectInput{
+		Bucket: &bucketName,
+		Key:    &keyName,
+		Body:   archive,
+	}
+
+	ctx, cancelFunc := context.WithTimeout(context.Background(), time.Minute)
+	defer cancelFunc()
+
+	_, err = client.PutObject(ctx, params)
+	return err
+}
+
+func ensureArtifactInBucket(bucketName string) error {
+	cfg, err := config.LoadDefaultConfig(context.TODO())
+	if err != nil {
+		return err
+	}
+
+	client := s3.NewFromConfig(cfg)
+
+	params := &s3.ListObjectsV2Input{
+		Bucket: &bucketName,
+	}
+
+	p := s3.NewListObjectsV2Paginator(client, params,
+		func(o *s3.ListObjectsV2PaginatorOptions) {
+			if v := int32(1); v != 0 {
+				o.Limit = v
+			}
```

Comment:
Not sure if I follow this part 🤔  
Isn't it essentially `o.Limit = int32(1)`?
Also is there a reason to set it to 1?

Reply:
Actually, you know, I copied the initialization of `p` from [an AWS example you can see here](https://github.com/aws/aws-sdk-go-v2/blob/main/example/service/s3/listObjects/listObjects.go#L58) and apparently didn't read very closely, because now that I am looking at it more carefully .... I don't know why it's done this way either? Maybe so I'll get caught copy-pasting code and not reading it closely enough :sweat_smile: 

What do you think? Should I just change it to `int32()` and assume the guy who wrote the AWS example was worried about different architectures that we don't support, or something?

Reply:
> Also is there a reason to set it to 1?

:shrug: it's what the docs said to do